### PR TITLE
Model functions now include metadata

### LINKED ISSUE #108: [Feature Request] Logging of custom training metrics
# 🚀 Feature Request

When training a model with `ModelTrainer`, it would be nice to be able to log some custom metrics (ideally in tensorboard), defined by the model (e.g., the values of the individual loss terms if the loss of the model is a sum of multiple terms). Right now one can only access the overall loss of the model.

## Motivation

**Is your feature request related to a problem? Please describe.**

At the moment I am working on a model that optimizes a sum of reconstruction loss, reward prediction loss, and a kl divergence term. For debugging purposes it would be nice to monitor how the individual losses evolve over time.
This logging can not be done by the model class on its own since it needs some information from the RL algorithm (e.g. the current iteration of the algorithm / number of samples drawn from the environment) for the logged values to be meaningful.

## Pitch

**Describe the solution you'd like**

The simplest solution certainly is to just allow passing kwargs to `ModelTrainer.train()`, which are passed through to `Model.update()`. This would allow to pass some custom logging function / object that then logs values passed by the model implementation.
This is of course not the most elegant solution, but the kwargs could also be used for other purposes (e.g. passing some additional information to `Model.update()` if a model implementation requires this).

**Describe alternatives you've considered**

An alternative to this would be to let `Model.update()` return a dictionary of metrics in addition to the loss. This dictionary could then be returned by `ModelTrainer.train()` or it could be processed by the callback passed to the function.
This would of course cause breaking changes since the method signature of `Model` would need to be changed.

**Are you willing to open a pull request?** (See [CONTRIBUTING](../../CONTRIBUTING.md))
Yes

### PR DESCRIPTION
**Pull request by luisenp**
## Types of changes

This pull request includes an API change where `Model.update()`, `Model.loss()` and `Model.eval_score()` now return an additional output with any metadata that is useful for logging (or anything else); for example, reconstructions, gradients, entropy, individual loss terms, etc..

This also adds a new callback to `ModelTrainer` that is called after every batch is computed, using the loss/score and metadata as inputs (the other callback is called after every epoch). 

This will eventually become a breaking change (in v0.2.0), but for now I added some extra logic to keep models on the old API functional.

**TODO before merge:** 
- [x] adding unit tests for the new callback. 
- [x] test how it works in #86 
----------------------------------------

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

Closes #108, and the rationale is explained there.

## How Has This Been Tested (if it applies)
Original unit test pass

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](../../CONTRIBUTING.md) document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by luisenp** [comment]
@eugenevinitsky @JanS97

File: mbrl/models/model_trainer.py [hunk 1]
```diff
@@ -71,6 +72,8 @@ def train(
         patience: Optional[int] = None,
         improvement_threshold: float = 0.01,
         callback: Optional[Callable] = None,
+        meta_includes_grad_norm: bool = False,
```
Comment (eugenevinitsky) [inline mbrl/models/model_trainer.py:75]: Hmm, it seems like this arg is getting passed around a lot. Would it make sense to create a set of default post-loss callbacks and include this one in it? That would prevent the number of arguments from ballooning

File: notebooks/fit_gaussian_mlp_ensemble_1d.ipynb [hunk 2]
```diff
@@ -215,4 +215,4 @@
  },
  "nbformat": 4,
  "nbformat_minor": 2
-}
+}
```
Comment (eugenevinitsky) [inline notebooks/fit_gaussian_mlp_ensemble_1d.ipynb:218]: nitpick: this doesn't seem intended

File: mbrl/models/model_trainer.py (hunk 1 above)
Comment (eugenevinitsky) [reply mbrl/models/model_trainer.py:75]: That is, this seems like a subset of meta-data you might want a model to return but this arg seems to be migrating out into a lot of classes in a way that might make refactoring more difficult in the future

**Review by eugenevinitsky — APPROVED** [review]
Looks good and confirmed that most of the examples I could find ran w/o error.

File: mbrl/models/model_trainer.py (hunk 1 above)
Comment (luisenp) [reply mbrl/models/model_trainer.py:75]: Right. And also the main purpose of this is to avoid computing this data if it's not really needed, since you have to aggregate over all model parameters. But, I do agree with your general comment about the argument being passed around a lot. 

On the other hand, very likely the cost of logging this is small compared to forward/backward, so maybe this is a case of over-optimization? We might just get rid of this arg altogether and just include grad norm every time? Or maybe turn it off via a model method, rather than via the model trainer.

File: notebooks/fit_gaussian_mlp_ensemble_1d.ipynb (hunk 2 above)
Comment (luisenp) [reply notebooks/fit_gaussian_mlp_ensemble_1d.ipynb:218]: I think this was the IDE's decision, but It's loading properly in Github right now so I'll leave it like this :P

File: mbrl/models/model_trainer.py (hunk 1 above)
Comment (eugenevinitsky) [reply mbrl/models/model_trainer.py:75]: Yeah, agree that I'm probably over-optimizing here. I think it probably makes since to always include it since it for standard tiny RL models the cost of computing will be tiny and cut down on extra args

Comment (luisenp) [reply mbrl/models/model_trainer.py:75]: Actually, I meant I was over-optimizing by including this arg to avoid computing the grad. I think this is a good suggestion, so I'll just remove the arg to keep the code cleaner, and consider some way of turning it off if it ever becomes an issue later.

Comment (luisenp) [reply mbrl/models/model_trainer.py]: Done.

File: mbrl/models/model_trainer.py [hunk 3]
```diff
@@ -117,14 +124,30 @@ def train(
...
+                    loss = cast(float, loss_and_maybe_meta[0])
+                    meta = cast(Dict, loss_and_maybe_meta[1])
+                else:
+                    # TODO remove this if in v0.2.0
+                    warnings.warn(_NO_META_WARNING_MSG)
+                    loss = cast(float, loss_and_maybe_meta)
+                    meta = None
                 batch_losses.append(loss)
+                if batch_callback_epoch:
+                    batch_callback_epoch(loss, meta)
```
Comment (JanS97) [inline mbrl/models/model_trainer.py:144]: I think it might make sense to pass the number of the current batch to the callback as well. E.g. if the epochs are really long, a user might want to log every x batches or just print some kind of progress bar with the number of the current batch. Of course one can keep some kind of counter inside the callback to count the batches, but that would be a bit ugly.
The same changes would also be necessary in `ModelTrainer.evaluate()`.

File: mbrl/models/model_trainer.py [hunk 4]
```diff
@@ -190,15 +218,25 @@ def evaluate(self, dataset: TransitionIterator) -> torch.Tensor:
...
+                batch_score = cast(torch.Tensor, batch_score_and_maybe_meta[0])
+                meta = cast(Dict, batch_score_and_maybe_meta[1])
+            else:
+                # TODO remove this "else" in v0.2.0
+                warnings.warn(_NO_META_WARNING_MSG)
+                batch_score = cast(torch.Tensor, batch_score_and_maybe_meta)
+                meta = None
+            batch_scores_list.append(batch_score)
+            if batch_callback:
+                batch_callback(batch_score.mean(), meta)
```
Comment (JanS97) [inline mbrl/models/model_trainer.py:232]: At the moment there is actually no indicator passed to the callback whether it is called from `evaluate()` or from `train()`. This makes it impossible to log training and evaluation metrics separately.
The callback should get some flag whether it is called from `evaluate()` or from `train()`. Or there should be separate callbacks for `train()` and `evaluate()`, but having three callbacks passed to `ModelTrainer` would probably be just more confusing and I would expect that the callbacks for training and evaluation typically should be very similar.

File: mbrl/models/model.py [hunk 5]
```diff
@@ -155,15 +172,34 @@ def update(
...
+        loss_and_maybe_meta = self.loss(model_in, target)
+        if isinstance(loss_and_maybe_meta, tuple):
+            # TODO - v0.2.0 remove this back-compatibility logic
+            loss = cast(torch.Tensor, loss_and_maybe_meta[0])
+            meta = cast(Optional[Dict[str, Any]], loss_and_maybe_meta[1])
+            loss.backward()
+
+            if meta is not None:
+                with torch.no_grad():
+                    grad_norm: torch.Tensor = 0  # type: ignore
```
Comment (JanS97) [inline mbrl/models/model.py:189]:
```suggestion
                    grad_norm: float = 0
```
Both here and in line 193 a float is assigned to `grad_norm` and never a tensor, so the type should be `float`.

**Review by JanS97 — CHANGES_REQUESTED** [review]

File: mbrl/models/basic_ensemble.py [hunk 6]
```diff
@@ -193,9 +193,12 @@ def loss(  # type: ignore
         self,
         model_ins: Sequence[torch.Tensor],
         targets: Optional[Sequence[torch.Tensor]] = None,
-    ) -> torch.Tensor:
+    ) -> Tuple[torch.Tensor, Optional[Dict[str, Any]]]:
```
[the remark this answers is not in the document]
Comment (luisenp) [reply mbrl/models/basic_ensemble.py:196]: Agreed, this is leftover from an earlier version where I was returning `None` in GaussianMLP.

File: mbrl/models/model.py [hunk 7]
```diff
@@ -155,15 +172,34 @@ def update(
...
+                        grad_norm += p.grad.data.norm(2).item() ** 2
+                    meta["grad_norm"] = grad_norm
+            optimizer.step()
+            return loss.item(), meta
+
+        else:
+            warnings.warn(_NO_META_WARNING_MSG)
+            loss_and_maybe_meta.backward()
+            optimizer.step()
+            return loss_and_maybe_meta.item()
```
[the remark this answers is not in the document]
Comment (luisenp) [reply mbrl/models/model.py:203]: IIRC, these were needed in early versions of MBRL-Lib because they received sequences of tensors instead, but they are indeed pointless now, good catch!

File: mbrl/models/model_trainer.py (hunk 3 above)
Comment (luisenp) [reply mbrl/models/model_trainer.py:144]: I think there is a balance, since having a lol of arguments also makes the API more rigid. My inclination is to err on the side of flexibility, even if it requires the user to do slightly more work. The epoch counter I think is necessary, since otherwise there is no way for the user to check which epoch it is just by looking at what the callback gets.

File: mbrl/models/model_trainer.py (hunk 4 above)
Comment (luisenp) [reply mbrl/models/model_trainer.py:232]: I actually had this extra arg in the last commit, but after playing a bit with PlaNet I ended up removing it and force pushing an amend without it; the rationale was that this signaling could be taken care by the meta data itself (and same reasoning as above for keeping the API cleaner). 

Here I'm less sure though, and I'm inclined to add it back. One benefit of including it is that it makes it very explicit, so it's easy to avoid some silent error, and less responsibility on users to implement the metadata correctly.

File: mbrl/models/model_trainer.py (hunk 3 above)
Comment (JanS97) [reply mbrl/models/model_trainer.py:144]: I think the batch index as additional argument would not be too critical. But I see your point. Especially since the batch index is not really needed in a typical use case (per episode logging should be sufficient) and there is the workaround with the counter for the special cases. So I am fine with not adding this argument.

File: mbrl/models/basic_ensemble.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (luisenp) [reply mbrl/models/basic_ensemble.py]: Fixed.

File: mbrl/models/model_trainer.py (hunk 4 above)
Comment (JanS97) [reply mbrl/models/model_trainer.py:232]: I would prefer having this argument. I think in all cases it is relevant whether the logged metrics were generated on training or evaluation data. So without the argument, in every case the user would have to do this bit of extra work of adding a flag to the dict, which could be avoided.

File: mbrl/models/basic_ensemble.py [hunk 8]
```diff
@@ -231,9 +244,20 @@ def eval_score(  # type: ignore
```
[the remark this answers is not in the document]
Comment (luisenp) [reply mbrl/models/basic_ensemble.py:244]: Not sure why GItHub is not showing this as "Outdate", but this is also [fixed](https://github.com/facebookresearch/mbrl-lib/blob/lep.update_returns_metadata/mbrl/models/basic_ensemble.py#L227) now.

File: mbrl/models/gaussian_mlp.py [hunk 9]
```diff
@@ -295,12 +295,14 @@ def loss(
         self,
         model_in: torch.Tensor,
         target: Optional[torch.Tensor] = None,
-    ) -> torch.Tensor:
+    ) -> Tuple[torch.Tensor, Optional[Dict[str, Any]]]:
```
[the remark this answers is not in the document]
Comment (luisenp) [reply mbrl/models/gaussian_mlp.py]: Fixed.

File: mbrl/models/model.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (luisenp) [reply mbrl/models/model.py:203]: Will remove in a separate PR/commit.

File: mbrl/models/model_trainer.py (hunk 4 above)
Comment (luisenp) [reply mbrl/models/model_trainer.py:232]: Agreed, I'll add it.

File: mbrl/models/model.py (hunk 5 above)
Comment (luisenp) [reply mbrl/models/model.py]: Fixed.

File: mbrl/models/model_trainer.py (hunk 4 above)
Comment (luisenp) [reply mbrl/models/model_trainer.py]: Added, and also a unit test.

**Review by JanS97 — APPROVED** [review]
